### ISSUE TITLE
Визуальный генератор страницы из имеющихся блоков

### ISSUE THREAD

**Issue by pavel06081991** [bem-site/bem-forum-content-ru#129]
Привет. Заранее извиняюсь за текстовую перенасыщенность вопросов и за, порой, футуристические вопросы. А также заранее благодарю за время, уделенное
на ответы, спасибо.

1) не знаю насколько это покажется удобным, но, думаю, по крайне мере это интересно. Представим, что мы разработали библиотеку блоков, реализовали для
каждого блока все его технологи(js, css, bemhml и т.д.). Все, библиотека готова. Вспоминая своё детство, помню, как любил играть в конструктор лего, из
кучи разных деталек создавать что-угодно, собирая все в одну конструкцию. Вопрос заключается в следующем, может у вас реализован или в планах следующий
механизм: открываем страницу в браузере, справа в панельке у нас отображаются визуально все имеющиеся блоки и мы, перетягивая их мышкой на страницу, создаем
наш сайт. Перетянув один из блоков на страницу, в октрывшемся окошке мы задаем все нужные параметры для блока. Точно также перетягиваем нужные элементы блока.
Затем нажимаем сохранить страницу и у нас автоматом генерится bemjson дерево нашей страницы. Разумеется, все это можно делать и вручную в bemjson файле, но
от визуального создания страницы, как игра в лего, чтоли получаешь какое-то удовольствие, как в детстве, играя в лего, ну и плюс bemjson генерировался бы автоматом и люди,
которые не особо сильны во всяких json и т.д. смогли бы делать свои странички не погружаясь в дебри bemjson, i-bem и т.д.

2) хочется сделать следующее приложение: загрузилась страница в браузер. Далее после каких-либо
пользовательских действий перезагружается какая-либо часть страницы, данные для которой подгружаются
ajax-ом. В этих пришедших данных находятся новые блоки, которых не было изначально на странице.
Скажите, есть ли у вас механизм, который бы позволил на клиенте отслеживать какие новые блоки нужно
вставить в страницу и который бы запрашивал с сервера для этих блоков css, js и bemhtml шаблоны,
возвращая все эти данные в одном бандле, например, в json формате, а на клиенте бы эти данные вставлялись
бы на страницу и только после этого бы вставлялся html новых блоков на страницу? В этом докладке https://events.yandex.ru/lib/talks/1413/
на 39:26 парень спрашивает именно то, о чем я написал. В ответе прозвучало про технологию, которая у вас есть с участием динамического сервера, но
говорится, что она не в опен сорсе. Скажите, за год что-нибудь изменилось в плане этой технологии динамической подгрузки всех технологий блока?

3) уже около месяца в свободное от работы время читаю/изучаю все технологии бэм, плюс пересмотрел
не один десяток видео с различных ваших конференций и теперь, имея хорошее представление, наконец
готов начать писать свой мини проект ради практики БЭМ, но увидев об анонсе bem-core 3, желание
начинать что-то кодить на текущей версии подостыло. Скажите, пожалуйста, когда планируете выпустить
версию 3?

4) уже ни раз в различных видео упоминалось о библиотеке bem-components и о том, что люди
пишут свои библиотеки, которые можно использовать в своих проектах. 
Скажите, а где можно скачать/увидеть сторонние, написанные не вами библиотеки блоков? Все старания
по поиску таковых в гугле с использованием ключевого запроса "БЭМ библиотеки блоков" ведут
сугубо на bem.info и, в частности, на страницы с библиотеками bem-core и bem-components.

**Comment by operatino** [comment]
@ilyar 
> Очень близко к описанным фантазиям есть проект SourceJS вот демо проект https://github.com/sourcejs/lego и видео https://www.youtube.com/watch?v=cefy_U5NU4o

> Насколько близко и планируется ли поддержать БЭМ-экосистему блоков призываю ответить @operatino

Мы планируем поддерживать совместимость с BEM блоками, это будет как один из примеров использования в рамках доп. расширений для [SourceJS](http://sourcejs.com) и [Design-lego](https://www.youtube.com/watch?v=cefy_U5NU4o).

Сами инструменты будут оставаться абстрагированными от конкретных технологий, без заточки конкретно под BEM, веб компоненты и тп. 

В данный момент Design-lego открыто доступен для всех, в альфа версии - https://github.com/sourcejs/lego. Ближайших планов на его дальнейшее развитие пока нету, инструмент работает на простых юзкейсах и служит больше примером возможностей работы с компонентами в будущем. Приветствую всех желающих поучаствовать в разработке, пишите на r@rhr.me, обсудим.

@vlastv
> В докладе "где не слышно", говориться о http://patternlab.io/ Atomic Design

В случае с pattern lab, там нужно самому (в коде) собирать компоненты, движек позволяет их ренделить и предоставляет навигацию. Это же самое в SourceJS можно сделать с любыми технологиями, более гибко, но с чуть более высоким порогом вхождения.

Оригинальные patternlab работает на PHP, но есть много имплементаций, в том числе и на NodeJS.

**Comment by zxqfox** [comment]
@operatino попытался влезть в sourcejs — не хватает примеров для быстрого старта, не очень понятно, какую задачу он решает и как его правильно использовать. Не думали над README поработать?

**Comment by pavelpower** [comment]
```
@ilyar Названия технологий и их апи настолько разные, что убивают многое. У них нет зависимостей других технологий, которые стоило бы запилить, без этого конфиги больших проектов выглядят как какашка (см. bem-components и проекты с хакатона). Технологии жестко завязаны файловую систему и имена технологий — оба не критично, но имхо это не правильно и тоже не дают красиво расти конфигам на больших проектах, и сам конфиг становится отдельным небольшим проектом. И это все не субъективно. Для 80% пользователей это так ;-)
```
@zxqfox я вот смотрю, очень часто enb-шники не используют автогенерацию конфигов. Может быть с этим было бы куда меньше. Но именно здаровенные конфиги меня пугают в enb

**Comment by zxqfox** [comment]
@pavelpower Большие — не значит плохие, но значит детальные. Думаешь, что enb-шники не пользуются https://github.com/bem/generator-bem-stub ? Сомневаюсь, что никто из них не пользуется ;-)

**Comment by operatino** [comment]
@zxqfox 
> @operatino попытался влезть в sourcejs — не хватает примеров для быстрого старта, не очень понятно, какую задачу он решает и как его правильно использовать. Не думали над README поработать?

Readme не достаточно, будут видео. Установив можно быстро понять что из себя это представляет, и сам sourcejs.com на этом же движке.

**Comment by zxqfox** [comment]
![image 1](https://cloud.githubusercontent.com/assets/677518/5775409/9cc73d1e-9d8a-11e4-9441-740cc9600a59.jpg)

**Comment by zxqfox** [comment]
![image](https://cloud.githubusercontent.com/assets/677518/5775444/f8b1cff4-9d8a-11e4-8ecf-c50049b3d9a5.png)

**Comment by ilyar** [comment]
@zxqfox для быстрого старта попробуй http://sourcejs.com/docs/base/ :

    npm install -g yo grunt-cli generator-sourcejs
    mkdir sourcejs-folder && cd $_
    node app
    open http://localhost:8080/

@operatino правильно ли я понимаю, что [SourceJS](https://github.com/sourcejs/Source) можно сравнить с [bem-site-engine](https://github.com/bem/bem-site-engine) т.е. они похожим образом решают одну задачу, а именно собирают интерактивную документацию из отдельных компонентов?

**Comment by operatino** [comment]
@ilyar Да, база одинаковая. Разница в том, что SourceJS не завязан конкретно на BEM и фокусируется больше на процессах разработки интерфейсов, через библиотеку компонентов - отсюда улучшенная навигация, с поиском, скриншотами компонентов, система плагинов для взаимодействия с командой на основе документации, АПИ примеров блоков. 

К тому же, мы рассмытриваем SourceJS как платформу для разработки и тестирования компонентов, используя [Clarify](http://sourcejs.com/docs/clarify/) и подход разработки прямо в документации.

Железная структура BEM позволяет сильно упростить работу с компонентами, в SourceJS мы больше боремся с не столь структурированными компонентами. Но, как я говорил ранее, поддерживать совместимость с BEM будем, так что будет возможность работать из SourceJS с BEM так же, как и с bem-site-engine + дополнительные возможности поверх.

На подходе так же авторизация, что бы персонализировать действия в библиотеке, настраивать фунециональность под разные роли (разработчик, дизайнер и тп).

**Comment by Guria** [comment]
@operatino Тоже давно смотрел на SourceJS. Как то мимо прошел релиз 0.4.0, сейчас изучу. Уже сейчас есть хоть какие-то варианты использования его с библиотекой бем блоков?

**Comment by operatino** [comment]
@Guria, варианты всегда были, вопрос во времени настройки. Пока не прорабатывал, но это на подходе, сейчас все силы на примеры направлю, без допила новых фич.

**Comment by operatino** [comment]
@tadatuta И кстати да, последних комментарией на сайте https://ru.bem.info/forum/issues/129/ нету, видимо баг. 

Я лично просто GH Issues использую.

**Comment by tadatuta** [comment]
@operatino @Guria похоже, что отображаются первые 30, завел баг, спасибо!

**Comment by Rahnar** [comment]
Мое мнение , какой смысл такого "Лего" - все сразу стану ленивыми и просто отпадет смысл в чем либо разбираться и как это работает. И все кругом станут верстальщиками и бэм разработчиками)))

**Comment by zxqfox** [comment]
Это потому что ты еще не устал руками это все писать ;-)

**Comment by sipayRT** [comment]
> Мое мнение , какой смысл такого "Лего" - все сразу стану ленивыми и просто отпадет смысл в чем либо разбираться и как это работает. И все кругом станут верстальщиками и бэм разработчиками)))

Да, этого нельзя допускать! Даешь повышение порога вхождения в БЭМ!!! :)
